clients/drcachesim/analysis_tool.h

```diff
@@ -168,20 +183,32 @@ class analysis_tool_t {
         return "";
     }
     /**
-     * Invoked once for each trace shard prior to calling parallel_shard_memref() for
-     * that shard, this allows a tool to create data local to a shard.  The \p
-     * shard_index is a unique identifier allowing shard data to be stored into a
-     * global table if desired (typically for aggregation use in print_results()).
-     * The \p worker_data is the return value of parallel_worker_init() for the
-     * worker thread who will exclusively operate on this shard.  The return value
-     * here will be passed to each invocation of parallel_shard_memref() for that
-     * same shard.
+     * \deprecated The parallel_shard_init_stream() is what is called by the analyzer;
+     * this function is only called if the default implementation of
+     * parallel_shard_init_stream() is left in place and it calls this version.
      */
     virtual void *
     parallel_shard_init(int shard_index, void *worker_data)
     {
         return nullptr;
     }
+    /**
+     * Invoked once for each trace shard prior to calling parallel_shard_memref() for
+     * that shard, this allows a tool to create data local to a shard.  The \p
+     * shard_index is a unique identifier allowing shard data to be stored into a global
+     * table if desired (typically for aggregation use in print_results()).  The \p
+     * worker_data is the return value of parallel_worker_init() for the worker thread
+     * who will exclusively operate on this shard.  The \p shard_stream allows tools to
+     * query details of the underlying trace shard during parallel operation; it is
+     * valid only until parallel_shard_exit() is called.  The return value here will be
+     * passed to each invocation of parallel_shard_memref() for that same shard.
+     */
+    virtual void *
+    parallel_shard_init_stream(int shard_index, void *worker_data,
+                               memref_stream_t *shard_stream)
```

Comment:
Xref #5720: the shard_stream interface can probably be used to get the shard_suffix (introduced in #5720) too.

Reply:
I should add to the PR description the other options we discussed in the past: adding more fields to memref_t (but have to append and then these common fields are at different offsets in each type; or prepend and break compat); adding fields to process_memref which seems more intrusive than expanding these init routines.  Having an interface seems most flexible for future features.

Reply:
Exposing memref_stream_t seems the best way forward. Yes, it'd be good to document those details for posterity.

Reply:
Added notes to memtrace_stream.h and will add to the commit description too.